src/fsharp/service/IncrementalBuild.fs

```diff
@@ -1506,188 +585,282 @@ type IncrementalBuilder(tcGlobals, frameworkTcImports, nonFrameworkAssemblyInput
             projectChecked.Trigger()
             return false
-        | Some _ -> 
-            return true
-      }
+        | Some _ ->
+            return true }
     
-    member builder.GetCheckResultsBeforeFileInProjectEvenIfStale filename: PartialCheckResults option  = 
-        let slotOfFile = builder.GetSlotOfFileName filename
-        let result = 
-            match slotOfFile with
-            | (*first file*) 0 -> GetScalarResult(initialTcAccNode, partialBuild)
-            | _ -> GetVectorResultBySlot(tcStatesNode, slotOfFile-1, partialBuild)  
-        
-        match result with
-        | Some (tcAcc, timestamp) -> Some (PartialCheckResults.Create (tcAcc, timestamp))
-        | _ -> None
-        
+    member _.GetCheckResultsBeforeFileInProjectEvenIfStale filename: PartialCheckResults option =
+        getSlot filename
+        |> tryCreatePartialCheckResultsBeforeSlot       
     
-    member builder.AreCheckResultsBeforeFileInProjectReady filename = 
-        let slotOfFile = builder.GetSlotOfFileName filename
-        let cache = TimeStampCache defaultTimeStamp
-        match slotOfFile with
-        | (*first file*) 0 -> IncrementalBuild.IsReady cache (Target(initialTcAccNode, None)) partialBuild 
-        | _ -> IncrementalBuild.IsReady cache (Target(tcStatesNode, Some (slotOfFile-1))) partialBuild  
+    member _.AreCheckResultsBeforeFileInProjectReady filename = 
+        match getSlot filename with
+        | 0 (* first file *) -> true
+        | slot -> isChecked (slot - 1)
         
-    member __.GetCheckResultsBeforeSlotInProject (ctok: CompilationThreadToken, slotOfFile) = 
+    member _.GetCheckResultsBeforeSlotInProject (ctok: CompilationThreadToken, slot) = 
       cancellable {
         let cache = TimeStampCache defaultTimeStamp
-        let! result = 
-          cancellable {
-            match slotOfFile with
-            | (*first file*) 0 -> 
-                let! build = IncrementalBuild.Eval cache ctok SavePartialBuild initialTcAccNode partialBuild
-                return GetScalarResult(initialTcAccNode, build)
-            | _ -> 
-                let! build = IncrementalBuild.EvalUpTo cache ctok SavePartialBuild (tcStatesNode, (slotOfFile-1)) partialBuild
-                return GetVectorResultBySlot(tcStatesNode, slotOfFile-1, build)  
-          }
-        
-        match result with
-        | Some (tcAcc, timestamp) -> return PartialCheckResults.Create (tcAcc, timestamp)
-        | None -> return! failwith "Build was not evaluated, expected the results to be ready after 'Eval' (GetCheckResultsBeforeSlotInProject)."
-      }
+        checkTimeStamps ctok cache
+
+        match tryCreatePartialCheckResultsBeforeSlot slot with
+        | Some results -> return results
+        | _ ->
+            let! tcAcc = priorTypeCheck ctok UseCache slot
+            match slot with
+            | 0 -> 
+                return PartialCheckResults.Create(tcAcc, defaultTimeStamp)
+            | _ ->
```

Comment:
What if `slot` is very small or large?

Reply:
`priorTypeCheck` will call `typeCheck` which will throw if the slot is out of range. See `checkSlot`.